# Exercise 4

**Match searches with bookings**

- For every search in the searches file, find out whether the search ended up in a booking or not (using the info in the bookings file). For instance, search and booking origin and destination should match.
- For the bookings file, origin and destination are the columns `dep_port` and `arr_port`, respectively.
- Generate a CSV file with the search data, and an additional field, containing 1 if the search ended up in a booking, and 0 otherwise.

In [2]:
import numpy as np
import pandas as pd
#import matplotlib.pyplot as plt

pd.options.display.max_columns = None
#pd.options.display.max_rows = 100

In [ ]:
columns_to_clean = ['Date', 'Seg1Departure', 'Seg1Arrival', 'Seg1Date', 
                 'Seg2Departure', 'Seg2Arrival', 'Seg2Date', 
                 'Seg3Departure', 'Seg3Arrival', 'Seg3Date', 
                 'Seg4Departure', 'Seg4Arrival', 'Seg4Date', 
                 'Seg5Departure', 'Seg5Arrival', 'Seg5Date', 
                 'Seg6Departure', 'Seg6Arrival', 'Seg6Date']

In [2]:
data_iterator = pd.read_csv("../challenge/searches.csv.bz2", 
                            chunksize=5*10**5, 
                            sep='^', 
                            compression='bz2',
                            #nrows=10**6,
                            low_memory=True,
                            dtype=str,
                            #dtype=dtypes,
                            #na_values=[np.nan],
                            #memory_map=True
                           )

searches = pd.DataFrame()
for i, data_chunk in enumerate(data_iterator):
    #data_chunk = searches_data_preprocessing(data_chunk, columns_to_clean)
    #data_chunk.reset_index(inplace=True)

    searches = searches.append(data_chunk)
    searches.drop_duplicates(inplace=True)
    print("Chunk: %d, size of chunk %d"%(i+1, searches.shape[0]))

Chunk: 1, size of chunk 358999
Chunk: 2, size of chunk 358999
Chunk: 3, size of chunk 359003
Chunk: 4, size of chunk 359003
Chunk: 5, size of chunk 359003
Chunk: 6, size of chunk 359003
Chunk: 7, size of chunk 359003
Chunk: 8, size of chunk 359003
Chunk: 9, size of chunk 359003
Chunk: 10, size of chunk 359003
Chunk: 11, size of chunk 359003
Chunk: 12, size of chunk 359003
Chunk: 13, size of chunk 359003
Chunk: 14, size of chunk 359003
Chunk: 15, size of chunk 359003
Chunk: 16, size of chunk 359003
Chunk: 17, size of chunk 359003
Chunk: 18, size of chunk 359003
Chunk: 19, size of chunk 359003
Chunk: 20, size of chunk 359003
Chunk: 21, size of chunk 359003
Chunk: 22, size of chunk 359003
Chunk: 23, size of chunk 359003
Chunk: 24, size of chunk 359003
Chunk: 25, size of chunk 359003
Chunk: 26, size of chunk 359003
Chunk: 27, size of chunk 359003
Chunk: 28, size of chunk 359003
Chunk: 29, size of chunk 359003
Chunk: 30, size of chunk 359003
Chunk: 31, size of chunk 359003
Chunk: 32, size o

In [3]:
searches.tail()

Date      Time TxnCode  \
1436000                                          2013-01-01  20:25:57     MPT   
1436001   2013-01-01,10:15:33,MPT,b0af35b31588dc4ab06d5c...       NaN     NaN   
1436002                                          2013-01-01  18:04:49     MPT   
1436009                                          2013-01-01  19:57:57     MPT   
20390197                                         2013-10-13  18:57:54     MTP   

                                  OfficeID Country Origin Destination  \
1436000   624d8c3ac0b3a7ca03e3c167e0f48327      DE    TXL         AUH   
1436001                                NaN     NaN    NaN         NaN   
1436002                               3561      US    ICT         SFO   
1436009   28d7a8c95e4db88589d3d35b66920e78      DE    FRA         BGW   
20390197  e41c9d833aa74600552f2ed688b67d81      AT    VIE          HA   

         RoundTrip NbSegments Seg1Departure Seg1Arrival    Seg1Date  \
1436000          1          2           TXL         AUH  2013-01-26   
1436001        NaN        NaN           NaN         NaN         NaN   
1436002          1          2           ICT         SFO  2013-08-02   
1436009          1          2           FRA         BGW  2013-02-26   
20390197       NaN        NaN           NaN         NaN         NaN   

         Seg1Carrier Seg1BookingCode Seg2Departure Seg2Arrival    Seg2Date  \
1436000           D2             NaN           AUH         TXL  2013-02-02   
1436001          NaN             NaN           NaN         NaN         NaN   
1436002          NaN             NaN           SFO         ICT  2013-08-09   
1436009          NaN             NaN           BGW         FRA  2013-04-08   
20390197         NaN             NaN           NaN         NaN         NaN   

         Seg2Carrier Seg2BookingCode Seg3Departure Seg3Arrival Seg3Date  \
1436000           D2             NaN           NaN         NaN      NaN   
1436001          NaN             NaN           NaN         NaN      NaN   
1436002          NaN             NaN           NaN         NaN      NaN   
1436009          NaN             NaN           NaN         NaN      NaN   
20390197         NaN             NaN           NaN         NaN      NaN   

         Seg3Carrier Seg3BookingCode Seg4Departure Seg4Arrival Seg4Date  \
1436000          NaN             NaN           NaN         NaN      NaN   
1436001          NaN             NaN           NaN         NaN      NaN   
1436002          NaN             NaN           NaN         NaN      NaN   
1436009          NaN             NaN           NaN         NaN      NaN   
20390197         NaN             NaN           NaN         NaN      NaN   

         Seg4Carrier Seg4BookingCode Seg5Departure Seg5Arrival Seg5Date  \
1436000          NaN             NaN           NaN         NaN      NaN   
1436001          NaN             NaN           NaN         NaN      NaN   
1436002          NaN             NaN           NaN         NaN      NaN   
1436009          NaN             NaN           NaN         NaN      NaN   
20390197         NaN             NaN           NaN         NaN      NaN   

         Seg5Carrier Seg5BookingCode Seg6Departure Seg6Arrival Seg6Date  \
1436000          NaN             NaN           NaN         NaN   1ASIWS   
1436001          NaN             NaN           NaN         NaN      NaN   
1436002          NaN             NaN           NaN         NaN      NaN   
1436009          NaN             NaN           NaN         NaN      NaN   
20390197         NaN             NaN           NaN         NaN      NaN   

         Seg6Carrier Seg6BookingCode    From  \
1436000            0               0       0   
1436001          NaN             NaN     NaN   
1436002          NaN             NaN  1ASIWS   
1436009         1ASI               0       0   
20390197         NaN             NaN     NaN   

                         IsPublishedForNeg                    IsFromInternet  \
1436000   d41d8cd98f00b204e9800998ecf8427e                  

In [4]:
searches.to_csv('../challenge/searches_no_dups.csv', sep='^', index=False)

In [11]:
def searches_data_preprocessing(df, columns_to_clean): 
    '''
    
    '''
    df.drop_duplicates(inplace=True)
    df.dropna(inplace=True, subset=['Date', 'Seg1Departure'])
    
    for column in columns_to_clean:
        df[column] = df[column].str.strip()
    
        if 'Date' not in column:
            df[column] = df[column].str.upper()
            # Para saber si hay lengths superiores a 3 sin que nos de error por los missing values    
            cell_lengths = df[column].map(lambda x: len(x) if not isinstance(x, float) else x)
            cell_lengths_notna = cell_lengths[cell_lengths.notna()]
            cell_lengths_sum = (cell_lengths_notna != 3.0).sum()
            if cell_lengths_sum != 0:
                print(f'Alert! There are `{column}` values with length different than 3. Dropping them..')

                mask = cell_lengths_notna != 3.0
                wrong_data = cell_lengths_notna[mask]
                print(wrong_data)
                df.drop(index=wrong_data.index, inplace=True)
    return df      

In [12]:
columns_to_clean = ['Date', 'Seg1Departure', 'Seg1Arrival', 'Seg1Date', 
                 'Seg2Departure', 'Seg2Arrival', 'Seg2Date', 
                 'Seg3Departure', 'Seg3Arrival', 'Seg3Date', 
                 'Seg4Departure', 'Seg4Arrival', 'Seg4Date', 
                 'Seg5Departure', 'Seg5Arrival', 'Seg5Date', 
                 'Seg6Departure', 'Seg6Arrival', 'Seg6Date',
                ]

data_iterator = pd.read_csv("../challenge/searches_no_dups.csv", 
                            chunksize=10**5, 
                            sep='^', 
                            #compression='bz2',
                            #nrows=10**6,
                            low_memory=False,
                            #dtype=str,
                            #dtype=dtypes,
                            #na_values=[np.nan],
                            #memory_map=True
                           )

searches = pd.DataFrame()
for i, data_chunk in enumerate(data_iterator):
    data_chunk = searches_data_preprocessing(data_chunk, columns_to_clean)
    #data_chunk.reset_index(inplace=True)
    searches = searches.append(data_chunk)
    #searches.drop_duplicates(inplace=True)
    print("Chunk: %d, size of chunk %d"%(i+1, searches.shape[0]))

Chunk: 1, size of chunk 100000
Chunk: 2, size of chunk 200000
Chunk: 3, size of chunk 300000
Chunk: 4, size of chunk 359002


In [13]:
searches.shape

(359002, 45)

In [10]:
searches.tail()

Date      Time TxnCode                          OfficeID  \
358998  2013-12-25  20:54:39     FXA  7f8c2d9dfe430c2c6b19298dd7e0ff96   
358999  2013-01-01  20:25:57     MPT  624d8c3ac0b3a7ca03e3c167e0f48327   
359001  2013-01-01  18:04:49     MPT                              3561   
359002  2013-01-01  19:57:57     MPT  28d7a8c95e4db88589d3d35b66920e78   
359003  2013-10-13  18:57:54     MTP  e41c9d833aa74600552f2ed688b67d81   

       Country Origin Destination  RoundTrip  NbSegments Seg1Departure  \
358998      US    LAX         TLV        1.0         5.0           LAX   
358999      DE    TXL         AUH        1.0         2.0           TXL   
359001      US    ICT         SFO        1.0         2.0           ICT   
359002      DE    FRA         BGW        1.0         2.0           FRA   
359003      AT    VIE          HA        NaN         NaN           NaN   

       Seg1Arrival    Seg1Date Seg1Carrier Seg1BookingCode Seg2Departure  \
358998         AMS  2014-02-02          UV               J           AMS   
358999         AUH  2013-01-26          D2             NaN           AUH   
359001         SFO  2013-08-02         NaN             NaN           SFO   
359002         BGW  2013-02-26         NaN             NaN           BGW   
359003         NaN         NaN         NaN             NaN           NaN   

       Seg2Arrival    Seg2Date Seg2Carrier Seg2BookingCode Seg3Departure  \
358998         TLV  2014-02-03          UV               J           TLV   
358999         TXL  2013-02-02          D2             NaN           NaN   
359001         ICT  2013-08-09         NaN             NaN           NaN   
359002         FRA  2013-04-08         NaN             NaN           NaN   
359003         NaN         NaN         NaN             NaN           NaN   

       Seg3Arrival    Seg3Date Seg3Carrier Seg3BookingCode Seg4Departure  \
358998         AMS  2014-02-17          UV               J           AMS   
358999         NaN         NaN         NaN             NaN           NaN   
359001         NaN         NaN         NaN             NaN           NaN   
359002         NaN         NaN         NaN             NaN           NaN   
359003         NaN         NaN         NaN             NaN           NaN   

       Seg4Arrival    Seg4Date Seg4Carrier Seg4BookingCode Seg5Departure  \
358998         ATL  2014-02-17          UV               J           ATL   
358999         NaN         NaN         NaN             NaN           NaN   
359001         NaN         NaN         NaN             NaN           NaN   
359002         NaN         NaN         NaN             NaN           NaN   
359003         NaN         NaN         NaN             NaN           NaN   

       Seg5Arrival    Seg5Date Seg5Carrier Seg5BookingCode Seg6Departure  \
358998         LAX  2014-02-17          UV               J           NaN   
358999         NaN         NaN         NaN             NaN           NaN   
359001         NaN         NaN         NaN             NaN           NaN   
359002         NaN         NaN         NaN             NaN           NaN   
359003         NaN         NaN         NaN             NaN           NaN   

       Seg6Arrival Seg6Date Seg6Carrier Seg6BookingCode    From  \
358998         NaN      NaN         NaN             NaN    1ASI   
358999         NaN   1ASIWS           0               0       0   
359001         NaN      NaN         NaN             NaN  1ASIWS   
359002         NaN      NaN        1ASI               0       0   
359003         NaN      NaN         NaN             NaN     NaN   

                       IsPublishedForNeg                    IsFromInternet  \
358998                                 0                                 0   
358999  d41d8cd98f00b204e9800998ecf8427e                               FRA   
359001                                 0                                 0   
359002                                 0  d41d8cd98f00b204e9800998ecf8427e   
359003                               NaN      

In [8]:
data_iterator = pd.read_csv("../challenge/bookings.csv.bz2", 
                            chunksize = 5*10**5, 
                            sep = '^', 
                            compression = 'bz2',
                            #nrows = 10**6,
                            usecols = ['dep_port', 
                                       'arr_port', 
                                       'brd_time           ',
                                       'cre_date           ',
                                      ],
                            low_memory=True,
                            dtype=str,
                           )

bookings = pd.DataFrame()
for i, data_chunk in enumerate(data_iterator):
    #data_chunk = bookings_data_preprocessing(data_chunk)
    #data_chunk.reset_index(inplace=True)
    #data_chunk['Booking'] = 1
    bookings = bookings.append(data_chunk)
    bookings.drop_duplicates(inplace=True)
    print("Chunk: %d, size of chunk %d"%(i+1, bookings.shape[0]))

Chunk: 1, size of chunk 354095
Chunk: 2, size of chunk 708879
Chunk: 3, size of chunk 708879
Chunk: 4, size of chunk 708879
Chunk: 5, size of chunk 708879
Chunk: 6, size of chunk 708879
Chunk: 7, size of chunk 708879
Chunk: 8, size of chunk 708879
Chunk: 9, size of chunk 708879
Chunk: 10, size of chunk 708879
Chunk: 11, size of chunk 708882
Chunk: 12, size of chunk 708882
Chunk: 13, size of chunk 708882
Chunk: 14, size of chunk 708882
Chunk: 15, size of chunk 708882
Chunk: 16, size of chunk 708882
Chunk: 17, size of chunk 708882
Chunk: 18, size of chunk 708882
Chunk: 19, size of chunk 708882
Chunk: 20, size of chunk 708882
Chunk: 21, size of chunk 708882


In [9]:
bookings.to_csv('../challenge/bookings_no_dups.csv', sep='^', index=False)

In [6]:
def bookings_data_preprocessing(df): 
    '''
    
    '''
    #df.drop_duplicates(inplace=True)
    df.dropna(inplace=True, subset=['cre_date           '])
    
    for column in df.columns:
        df[column] = df[column].str.strip()
    
    
        if 'port' in column:
            df[column] = df[column].str.upper()
            
            # Para saber si hay lengths superiores a 3 sin que nos de error por los missing values
            cell_lengths = df[column].map(lambda x: len(x) if not isinstance(x, float) else x)
            cell_lengths_notna = cell_lengths[cell_lengths.notna()]
            cell_lengths_sum = (cell_lengths_notna != 3.0).sum()
            if cell_lengths_sum != 0:
                print(f'Alert! There are `{column}` values with length different than 3. Dropping them..')

                mask = cell_lengths_notna != 3.0
                wrong_data = cell_lengths_notna[mask]
                print(wrong_data)
                df.drop(index=wrong_data.index, inplace=True)
        else:
            df[column] = df[column].str.split().str[0]
   #df.dropna(inplace=True, subset=['cre_date'])
        
    return df      

In [7]:
data_iterator = pd.read_csv("../challenge/bookings_no_dups.csv", 
                            chunksize = 10**5, 
                            sep = '^', 
                            #compression = 'bz2',
                            #nrows = 10**6,
                            usecols = ['dep_port', 
                                       'arr_port', 
                                       'brd_time           ',
                                       'cre_date           ',
                                      ],
                            #low_memory=False,
                            #parse_dates=['cre_date           '],
                           )

bookings = pd.DataFrame()
for i, data_chunk in enumerate(data_iterator):
    data_chunk = bookings_data_preprocessing(data_chunk)
    #data_chunk.reset_index(inplace=True)
    data_chunk['Booking'] = 1
    bookings = bookings.append(data_chunk)
    print("Chunk: %d, size of chunk %d"%(i+1, bookings.shape[0]))


Chunk: 1, size of chunk 100000
Chunk: 2, size of chunk 200000
Chunk: 3, size of chunk 300000
Chunk: 4, size of chunk 400000
Chunk: 5, size of chunk 500000
Chunk: 6, size of chunk 600000
Chunk: 7, size of chunk 700000
Alert! There are `dep_port` values with length different than 3. Dropping them..
708880    2
Name: dep_port, dtype: int64
Chunk: 8, size of chunk 708881


In [8]:
bookings.shape

(708881, 5)

In [9]:
bookings.tail()

cre_date            dep_port arr_port brd_time             Booking
708876           2013-05-28      AUS      RDU          2013-07-12        1
708877           2013-05-28      TLS      ORY          2013-06-04        1
708878           2013-05-28      TLS      ORY          2013-06-04        1
708879           2013-03-26      AKL      SVO          2013-04-24        1
708881  2013-03-25,00:00:00      TYO      SIN          2013-04-16        1

Comprobar de nuevo los duplicados y eliminarlos

In [16]:
bookings.duplicated().sum()

49882

In [17]:
searches.duplicated().sum()

0

In [18]:
bookings.drop_duplicates(inplace=True)
bookings.duplicated().sum()

0

In [19]:
bookings.tail()

cre_date            dep_port arr_port brd_time             Booking
708874           2013-05-14      CMB      MAA          2013-05-24        1
708875           2013-05-14      CMB      MAA          2013-05-19        1
708876           2013-05-28      AUS      RDU          2013-07-12        1
708877           2013-05-28      TLS      ORY          2013-06-04        1
708881  2013-03-25,00:00:00      TYO      SIN          2013-04-16        1

In [20]:
print(bookings.shape)
print(searches.shape)

(658999, 5)
(359003, 45)


Merge

In [21]:
new_columns_bookings = [['Date', 'Seg1Departure', 'Seg1Arrival', 'Seg1Date', 'booking'], 
                        ['Date', 'Seg2Departure', 'Seg2Arrival', 'Seg2Date', 'booking'],
                        ['Date', 'Seg3Departure', 'Seg3Arrival', 'Seg3Date', 'booking'],
                        ['Date', 'Seg4Departure', 'Seg4Arrival', 'Seg4Date', 'booking'],
                        ['Date', 'Seg5Departure', 'Seg5Arrival', 'Seg5Date', 'booking'],
                        ['Date', 'Seg6Departure', 'Seg6Arrival', 'Seg6Date', 'booking'],
                       ]

In [22]:
for i, lista in enumerate(new_columns_bookings):
    bookings.columns = lista

    searches = searches.merge(bookings, 
                              how = 'left', 
                              on = lista[:4], 
                              suffixes=(i, i+1),
                             )
    print(searches.shape)

(359003, 46)
(359003, 47)
(359003, 48)
(359003, 49)
(359003, 50)
(359003, 51)


In [23]:
searches.iloc[:, -6:].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 359003 entries, 0 to 359002
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   booking1  451 non-null    float64
 1   booking2  181 non-null    float64
 2   booking3  31 non-null     float64
 3   booking4  25 non-null     float64
 4   booking5  3 non-null      float64
 5   booking6  1 non-null      float64
dtypes: float64(6)
memory usage: 19.2 MB


In [24]:
searches.iloc[:, -6:] = searches.iloc[:, -6:].fillna(0)

In [25]:
searches.iloc[:, -6:].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 359003 entries, 0 to 359002
Data columns (total 6 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   booking1  359003 non-null  float64
 1   booking2  359003 non-null  float64
 2   booking3  359003 non-null  float64
 3   booking4  359003 non-null  float64
 4   booking5  359003 non-null  float64
 5   booking6  359003 non-null  float64
dtypes: float64(6)
memory usage: 19.2 MB


In [26]:
searches.iloc[:, -6:].sum()

booking1    451.0
booking2    181.0
booking3     31.0
booking4     25.0
booking5      3.0
booking6      1.0
dtype: float64

Creamos la nueva columna

In [27]:
searches['Booking'] = 0
for column in searches.columns[-7:-1]:
    searches['Booking'] += searches[column]

In [28]:
searches['Booking'].sum()

692.0

In [29]:
searches.iloc[:, -7:-1].sum().sum() #comprobacion

692.0

Eliminamos las columnas sobrantes

In [30]:
searches.columns[-7:-1]

Index(['booking1', 'booking2', 'booking3', 'booking4', 'booking5', 'booking6'], dtype='object')

In [31]:
print(searches.shape)
searches.drop(columns=searches.columns[-7:-1], inplace=True)
print(searches.shape)

(359003, 52)
(359003, 46)


In [32]:
searches['Booking'].sum()

692.0

In [27]:
bookings.isna().sum()

cre_date               0
dep_port               0
arr_port               0
brd_time               0
Booking                0
dtype: int64

In [28]:
searches.isna().sum()

Date                      0
Time                      0
TxnCode                   0
OfficeID                  0
Country                  51
Origin                    0
Destination               0
RoundTrip                 0
NbSegments                0
Seg1Departure             0
Seg1Arrival               0
Seg1Date               1282
Seg1Carrier          284543
Seg1BookingCode      411897
Seg2Departure        132239
Seg2Arrival          132239
Seg2Date             133765
Seg2Carrier          337279
Seg2BookingCode      418449
Seg3Departure        434559
Seg3Arrival          434559
Seg3Date             434691
Seg3Carrier          435786
Seg3BookingCode      437078
Seg4Departure        439480
Seg4Arrival          439480
Seg4Date             439599
Seg4Carrier          439648
Seg4BookingCode      439945
Seg5Departure        455221
Seg5Arrival          455221
Seg5Date             455249
Seg5Carrier          455244
Seg5BookingCode      455309
Seg6Departure        456732
Seg6Arrival         

|Column|NaN|
|:---:|:---:|
|Date|0|
|Seg1Departure|0|
|Seg1Arrival|0|
|Seg1Date|1282|

|Column|NaN|
|:---:|:---:|
|Date|0|
|Seg2Departure|132239|
|Seg2Arrival|132239|
|Seg2Date|133765|

Pdemos observar el mismo patron en todos los segmentos, `SegnDeparture` y `SegnArribal` tienen la misma cantidad de missing values, mientras que `SegnDate` tiene unos pocos mas. 

Ello significa que si un cierto dia se reservan vuelos de una ciudad a otra en dias distintos, estos tendran los mismos datos en las columnas relevantes para hacer el join.

Esto lo vamos a dejar para el finde.